### PR TITLE
CORDA-3083: Backport dumpCheckpoints shell command

### PR DESCRIPTION
**Pull request by lankydan**
Port of #5171.

### CODE REVIEW

File: docs/source/upgrading-cordapps.rst [hunk 1]
```diff
@@ -275,6 +275,17 @@ a drain is complete there should be no outstanding checkpoints or running flows.
 A node can be drained or undrained via RPC using the ``setFlowsDrainingModeEnabled`` method, and via the shell using
 the standard ``run`` command to invoke the RPC. See :doc:`shell` to learn more.
 
+To assist in draining a node, the ``dumpCheckpoints`` shell command will output JSON representations of each checkpointed flow.
+A zip containing the JSON files is created in the ``logs`` directory of the node. This information can then be used to determine the
+state of stuck flows or flows that experienced internal errors and were kept in the node for manual intervention. To drain these flows,
+the node will need to be restarted or the flow will need to be removed using ``killFlow``.
```
Comment (shamsasari) [inline docs/source/upgrading-cordapps.rst]: V3 does not have the `killFlow` RPC.

File: node/src/main/kotlin/net/corda/node/services/api/CheckpointStorage.kt [hunk 2]
```diff
@@ -27,6 +28,7 @@ interface CheckpointStorage {
      */
     fun forEach(block: (Checkpoint) -> Boolean)
 
+    fun getAllCheckpoints(): Stream<Pair<String, Checkpoint>>
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/api/CheckpointStorage.kt]: Add a comment to this method that was ported for the dumpCheckpoints and that it's a replacement for `forEach` above.

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+            } else {
+                log.info("Flow dump already in progress, skipping current call")
+            }
+        } finally {
+            lock.decrementAndGet()
+        }
+    }
+
+    private fun FlowStateMachineImpl<*>.toJson(id: String, now: Instant): CheckpointJson {
+        val (fiber, flowLogic) = this to this.logic
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: I think you're trying to keep the code similar to V4, but I think it's better to keep to the natural coding style and stick to `this` and `logic`.

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+        var secondsSpentWaiting: Long = 0
+    }
+
+    @Suppress("unused")
+    private class SendAndReceiveJson(val session: FlowSession, val sentPayloadType: Class<*>?, val sentPayload: Any?, val receivedPayloadType: Class<*>?)
+
+    private fun WaitingRequest.toSuspendedOn(suspendedTimestamp: Instant, now: Instant): SuspendedOn {
+        fun SendAndReceive.toJson(): SendAndReceiveJson {
+            val payload = when (this.message) {
+                is ExistingSessionMessage -> (message.payload as? DataSessionMessage)?.payload?.deserialize()
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: If the payload is undeserialisable for whatever reason then we prevent the checkpoint from being viewable. Best to catch the exception and replace the payload with some message Sytinh like `"*** Unable to deserialise message payload ***"`

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+            val attachments: List<SecureHash>,
+            val privacySalt: PrivacySalt
+    )
+
+    @JsonIgnoreProperties("flow", "fiber")
+    @JsonAutoDetect(getterVisibility = Visibility.NONE, isGetterVisibility = Visibility.NONE, fieldVisibility = Visibility.ANY)
+    private interface FlowSessionInternalMixin
+}
+
+// taken from newer corda version due to parent classes not being searched for fields
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: Capital 'T' in taken

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt (hunk 3 above)
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: ughhh i forgot to remove that line...

**Comment by lankydan** [comment]
Example output

```json
{
  "flowId" : "D6C7319C1AEA785113ECC24F2F98655502C46BA799F2D4A0A01D6A02D2A00F42",
  "topLevelFlowClass" : "com.lankydanblog.tutorial.base.flows.SendMessageFlow",
  "flowCallStack" : [
    {
      "flowClass" : "com.lankydanblog.tutorial.base.flows.SendMessageFlow",
      "progressStep" : null,
      "flowLogic" : {
        "message" : {
          "sender" : "O=partyB, L=London, C=GB",
          "recipient" : "O=partyA, L=London, C=GB",
          "contents" : "1",
          "linearId" : {
            "externalId" : "abfd2587-a0e9-485d-824a-b1df0a73d183",
            "id" : "496bbfc0-07f7-483d-8b79-797de2d781d4"
          },
          "participants" : [
            "O=partyB, L=London, C=GB",
            "O=partyA, L=London, C=GB"
          ]
        }
      }
    },
    {
      "flowClass" : "net.corda.core.flows.CollectSignaturesFlow",
      "progressStep" : "Collecting signatures from counterparties.",
      "flowLogic" : {
        "partiallySignedTx" : {
          "wire" : {
            "id" : "9649C8FC86F7AD8EAB23BC3DF715E017CD0B2D4C81CA2C2628C5C51DA2BC51FB",
            "notary" : "O=Notary, L=London, C=GB",
            "inputs" : [ ],
            "outputs" : [
              {
                "data" : {
                  "sender" : "O=partyB, L=London, C=GB",
                  "recipient" : "O=partyA, L=London, C=GB",
                  "contents" : "1",
                  "linearId" : {
                    "externalId" : "abfd2587-a0e9-485d-824a-b1df0a73d183",
                    "id" : "496bbfc0-07f7-483d-8b79-797de2d781d4"
                  },
                  "participants" : [
                    "O=partyB, L=London, C=GB",
                    "O=partyA, L=London, C=GB"
                  ]
                },
                "contract" : "com.lankydanblog.tutorial.contracts.MessageContract",
                "notary" : "O=Notary, L=London, C=GB",
                "encumbrance" : null,
                "constraint" : { }
              }
            ],
            "commands" : [
              {
                "value" : { },
                "signers" : [
                  "GfHq2tTVk9z4eXgyWRuTexHYezFZRJjwrEkAXkYEXtE9qkRNLFV7xxZSkS22",
                  "GfHq2tTVk9z4eXgyPE9bmVyDnpA7SLbE7dhrN4bUKPQPYNtBVvMqrUdajUyv"
                ]
              }
            ],
            "timeWindow" : null,
            "attachments" : [
              "12C837CE2688A82E20CCB00A31D52655AFBDDD8AB13B621401F8A86D8066C5D5"
            ],
            "privacySalt" : "dp+akmj3xa6SfEFQ7UHhYV33MUMkDrPX18pu+PRTYdI="
          },
          "signatures" : [
            {
              "bytes" : "PJokogXEsC++SvEbjCCf2EmpIBqlAkttmlhxOKj7Q+0bOY8DywvSO2kk++FjeMUzsmp+TpzmULXRpwvyLZOyBQ==",
              "by" : "GfHq2tTVk9z4eXgyWRuTexHYezFZRJjwrEkAXkYEXtE9qkRNLFV7xxZSkS22",
              "signatureMetadata" : {
                "platformVersion" : 3,
                "schemeNumberID" : 4
              },
              "partialMerkleTree" : null
            }
          ]
        },
        "sessionsToCollectFrom" : [
          {
            "peer" : "O=partyA, L=London, C=GB"
          }
        ],
        "myOptionalKeys" : null
      }
    },
    {
      "flowClass" : "net.corda.core.flows.CollectSignatureFlow",
      "progressStep" : null,
      "flowLogic" : {
        "partiallySignedTx" : {
          "wire" : {
            "id" : "9649C8FC86F7AD8EAB23BC3DF715E017CD0B2D4C81CA2C2628C5C51DA2BC51FB",
            "notary" : "O=Notary, L=London, C=GB",
            "inputs" : [ ],
            "outputs" : [
              {
                "data" : {
                  "sender" : "O=partyB, L=London, C=GB",
                  "recipient" : "O=partyA, L=London, C=GB",
                  "contents" : "1",
                  "linearId" : {
                    "externalId" : "abfd2587-a0e9-485d-824a-b1df0a73d183",
                    "id" : "496bbfc0-07f7-483d-8b79-797de2d781d4"
                  },
                  "participants" : [
                    "O=partyB, L=London, C=GB",
                    "O=partyA, L=London, C=GB"
                  ]
                },
                "contract" : "com.lankydanblog.tutorial.contracts.MessageContract",
                "notary" : "O=Notary, L=London, C=GB",
                "encumbrance" : null,
                "constraint" : { }
              }
            ],
            "commands" : [
              {
                "value" : { },
                "signers" : [
                  "GfHq2tTVk9z4eXgyWRuTexHYezFZRJjwrEkAXkYEXtE9qkRNLFV7xxZSkS22",
                  "GfHq2tTVk9z4eXgyPE9bmVyDnpA7SLbE7dhrN4bUKPQPYNtBVvMqrUdajUyv"
                ]
              }
            ],
            "timeWindow" : null,
            "attachments" : [
              "12C837CE2688A82E20CCB00A31D52655AFBDDD8AB13B621401F8A86D8066C5D5"
            ],
            "privacySalt" : "dp+akmj3xa6SfEFQ7UHhYV33MUMkDrPX18pu+PRTYdI="
          },
          "signatures" : [
            {
              "bytes" : "PJokogXEsC++SvEbjCCf2EmpIBqlAkttmlhxOKj7Q+0bOY8DywvSO2kk++FjeMUzsmp+TpzmULXRpwvyLZOyBQ==",
              "by" : "GfHq2tTVk9z4eXgyWRuTexHYezFZRJjwrEkAXkYEXtE9qkRNLFV7xxZSkS22",
              "signatureMetadata" : {
                "platformVersion" : 3,
                "schemeNumberID" : 4
              },
              "partialMerkleTree" : null
            }
          ]
        },
        "session" : {
          "peer" : "O=partyA, L=London, C=GB"
        },
        "signingKeys" : [
          "GfHq2tTVk9z4eXgyPE9bmVyDnpA7SLbE7dhrN4bUKPQPYNtBVvMqrUdajUyv"
        ]
      }
    },
    {
      "flowClass" : "net.corda.core.flows.SendTransactionFlow",
      "progressStep" : null,
      "flowLogic" : {
        "otherSideSession" : {
          "peer" : "O=partyA, L=London, C=GB"
        },
        "payload" : {
          "wire" : {
            "id" : "9649C8FC86F7AD8EAB23BC3DF715E017CD0B2D4C81CA2C2628C5C51DA2BC51FB",
            "notary" : "O=Notary, L=London, C=GB",
            "inputs" : [ ],
            "outputs" : [
              {
                "data" : {
                  "sender" : "O=partyB, L=London, C=GB",
                  "recipient" : "O=partyA, L=London, C=GB",
                  "contents" : "1",
                  "linearId" : {
                    "externalId" : "abfd2587-a0e9-485d-824a-b1df0a73d183",
                    "id" : "496bbfc0-07f7-483d-8b79-797de2d781d4"
                  },
                  "participants" : [
                    "O=partyB, L=London, C=GB",
                    "O=partyA, L=London, C=GB"
                  ]
                },
                "contract" : "com.lankydanblog.tutorial.contracts.MessageContract",
                "notary" : "O=Notary, L=London, C=GB",
                "encumbrance" : null,
                "constraint" : { }
              }
            ],
            "commands" : [
              {
                "value" : { },
                "signers" : [
                  "GfHq2tTVk9z4eXgyWRuTexHYezFZRJjwrEkAXkYEXtE9qkRNLFV7xxZSkS22",
                  "GfHq2tTVk9z4eXgyPE9bmVyDnpA7SLbE7dhrN4bUKPQPYNtBVvMqrUdajUyv"
                ]
              }
            ],
            "timeWindow" : null,
            "attachments" : [
              "12C837CE2688A82E20CCB00A31D52655AFBDDD8AB13B621401F8A86D8066C5D5"
            ],
            "privacySalt" : "dp+akmj3xa6SfEFQ7UHhYV33MUMkDrPX18pu+PRTYdI="
          },
          "signatures" : [
            {
              "bytes" : "PJokogXEsC++SvEbjCCf2EmpIBqlAkttmlhxOKj7Q+0bOY8DywvSO2kk++FjeMUzsmp+TpzmULXRpwvyLZOyBQ==",
              "by" : "GfHq2tTVk9z4eXgyWRuTexHYezFZRJjwrEkAXkYEXtE9qkRNLFV7xxZSkS22",
              "signatureMetadata" : {
                "platformVersion" : 3,
                "schemeNumberID" : 4
              },
              "partialMerkleTree" : null
            }
          ]
        }
      }
    }
  ],
  "suspendedOn" : {
    "receive" : {
      "session" : {
        "flowSession" : {
          "peer" : "O=partyA, L=London, C=GB"
        },
        "ourSessionId" : -902223753937324007,
        "initiatingParty" : null,
        "state" : {
          "otherParty" : "O=partyA, L=London, C=GB",
          "sendToParty" : "O=partyA, L=London, C=GB"
        },
        "retryable" : false,
        "receivedMessages" : [ ]
      },
      "userReceiveType" : null,
      "stackTraceInCaseOfProblems" : null
    },
    "suspendedTimestamp" : "2019-07-17T13:20:41",
    "secondsSpentWaiting" : 109
  },
  "origin" : {
    "rpc" : "user1"
  },
  "ourIdentity" : "O=partyB, L=London, C=GB",
  "activeSessions" : [
    {
      "peer" : "O=partyA, L=London, C=GB",
      "ourSessionId" : -902223753937324007,
      "receivedMessages" : [ ],
      "peerSessionId" : null
    }
  ]
}
```

File: docs/source/upgrading-cordapps.rst [hunk 6]
```diff
@@ -275,6 +275,17 @@ a drain is complete there should be no outstanding checkpoints or running flows.
 A node can be drained or undrained via RPC using the ``setFlowsDrainingModeEnabled`` method, and via the shell using
 the standard ``run`` command to invoke the RPC. See :doc:`shell` to learn more.
 
+To assist in draining a node, the ``dumpCheckpoints`` shell command will output JSON representations of each checkpointed flow.
+A zip containing the JSON files is created in the ``logs`` directory of the node. This information can then be used to determine the
+state of stuck flows or flows that experienced internal errors and were kept in the node for manual intervention. To drain these flows,
+the node will need to be restarted or the flow will need to be removed manually.
+
+.. warning:: Deleting checkpoints manually or via ``killFlow`` can lead to an inconsistent ledger among transacting parties. Great care
```
Comment (shamsasari) [inline docs/source/upgrading-cordapps.rst]: Still referencing `killFlow`

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+                                    try {
+                                        val jsonBytes = writer.writeValueAsBytes(json)
+                                        zip.putNextEntry(ZipEntry("${json.topLevelFlowClass.simpleName}-$runId.json"))
+                                        zip.write(jsonBytes)
+                                        zip.closeEntry()
+                                    } catch (e: Exception) {
+                                        log.info("Failed to output json checkpoint dump for ${json.topLevelFlowClass.simpleName}-$runId", e)
+                                    }
+                                } catch (e: Exception) {
+                                    log.info("Failed to deserialise checkpoint with id: $runId")
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: The exception should be included as the second parameter.

Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: There should also be an entry for this flow in the dump with a similar message.

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+                                    val checkpoint = serialisedCheckpoint.serializedFiber.deserialize(
+                                            context = SerializationDefaults.CHECKPOINT_CONTEXT.withTokenContext(checkpointSerializationContext)
+                                    )
+                                    val json = checkpoint.toJson(runId, now)
+                                    try {
+                                        val jsonBytes = writer.writeValueAsBytes(json)
+                                        zip.putNextEntry(ZipEntry("${json.topLevelFlowClass.simpleName}-$runId.json"))
+                                        zip.write(jsonBytes)
+                                        zip.closeEntry()
+                                    } catch (e: Exception) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: I'm not sure we need to worry about this error

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+        }.also {
+            it.suspendedTimestamp = suspendedTimestamp
+            it.secondsSpentWaiting = TimeUnit.MILLISECONDS.toSeconds(Duration.between(suspendedTimestamp, now).toMillis())
+        }
+    }
+
+    private fun SerializedBytes<Any>.deserializeOrOutputPlaceholder() = try {
+        deserialize()
+    } catch (e: Exception) {
+        "*** Unable to deserialise message payload ***"
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: I think the exception message would be useful as well.

File: node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+                                try {
+                                    val checkpoint = serialisedCheckpoint.serializedFiber.deserialize(
+                                            context = SerializationDefaults.CHECKPOINT_CONTEXT.withTokenContext(checkpointSerializationContext)
+                                    )
+                                    val json = checkpoint.toJson(runId, now)
+                                    val jsonBytes = writer.writeValueAsBytes(json)
+                                    zip.putNextEntry(ZipEntry("${json.topLevelFlowClass.simpleName}-$runId.json"))
+                                    zip.write(jsonBytes)
+                                    zip.closeEntry()
+                                } catch (e: Exception) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: Technically we could reduce the scope of the try-catch to just the `deserialize` call, but that's minor so this is OK for now.

Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: I wanted it to catch jackson errors as well. I have tested it the best I can, but I think there are some classes I haven't managed to test properly. 

With this `try/catch` jackson errors are caught and will at least let everything else continue. Worst case, we can address these at another time if a number of jackson errors occur.

Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumper.kt]: Yes. I guess it's the zip logic that doesn't need to be in the try block. But I'm being picky so ignore me :).

**Comment by lankydan** [comment]
All unit tests pass locally

**Pull request merged**